crawl-ref/source/util/species-gen.py

```diff
@@ -0,0 +1,682 @@
+#!/usr/bin/env python3
+
+import argparse
+import os
+import sys
+import traceback
+import collections
+
+import yaml  # pip install pyyaml
+
+
+class Species(dict):
+    """Parser for YAML definition files.
+
+    If any YAML content is invalid, the relevant parser function below should
+    raise ValueError.
+    """
+    @staticmethod
+    def from_yaml(s):
+        species = Species()
+        species['tag_major_version_opener'] = '' if 'TAG_MAJOR_VERSION' not in s else "#if TAG_MAJOR_VERSION == %s" % s['TAG_MAJOR_VERSION']
+        species['tag_major_version_closer'] = '' if 'TAG_MAJOR_VERSION' not in s else "#endif"
+        species['enum'] = s.get('enum', "SP_%s" % s['name'].replace(' ', '_').upper())
+        species['name'] = s['name']
+        species['short_name'] = s.get('short_name', s['name'][:2])
+        species['adjective'] = quote(s['adjective']) if 'adjective' in s else 'nullptr'
+        species['genus'] = quote(s['genus']) if 'genus' in s else 'nullptr'
+        species['species_flags'] = species_flags(s.get('species_flags', []))
+        species['xp'] = s['aptitudes']['xp']
+        species['hp'] = s['aptitudes']['hp']
+        species['mp'] = s['aptitudes']['mp_mod']
+        species['mr'] = s['aptitudes']['mr']
+        species['aptitudes'] = aptitudes(s['aptitudes'])
+        species['monster_name'] = s.get('monster', "MONS_%s" % s['name'].replace(' ', '_').upper())
+        species['habitat'] = 'HT_LAND' if not s.get('can_swim') else 'HT_WATER'
+        species['undead'] = s.get('undead_type', 'US_ALIVE')
```

Comment:
There doesn't appear to be validation here, and the `US_` prefix has to be included; I'd suggest implementing and using `enumify()`, probably with an optional list of permissible enums for validation.

Reply:
Added `enumify` function as suggested, although validation of acceptable values is kept to field-specific functions.